internal/cluster/cmd/clusterd/contract.go

```diff
@@ -0,0 +1,489 @@
+package main
+
+import (
+	"bytes"
+	"math/bits"
+	"sort"
+	"time"
+
+	"go.sia.tech/core/consensus"
+	proto2 "go.sia.tech/core/rhp/v2"
+	"go.sia.tech/core/types"
+	"go.sia.tech/coreutils"
+	"go.sia.tech/coreutils/chain"
+	"go.uber.org/zap"
+)
+
+// NOTE: due to a bug in the transaction validation code, calculating payouts
+// is way harder than it needs to be. Tax is calculated on the post-tax
+// contract payout (instead of the sum of the renter and host payouts). So the
+// equation for the payout is:
+//
+//	   payout = renterPayout + hostPayout + payout*tax
+//	∴  payout = (renterPayout + hostPayout) / (1 - tax)
+//
+// This would work if 'tax' were a simple fraction, but because the tax must
+// be evenly distributed among siafund holders, 'tax' is actually a function
+// that multiplies by a fraction and then rounds down to the nearest multiple
+// of the siafund count. Thus, when inverting the function, we have to make an
+// initial guess and then fix the rounding error.
+func taxAdjustedPayout(target types.Currency) types.Currency {
+	// compute initial guess as target * (1 / 1-tax); since this does not take
+	// the siafund rounding into account, the guess will be up to
+	// types.SiafundCount greater than the actual payout value.
+	guess := target.Mul64(1000).Div64(961)
+
+	// now, adjust the guess to remove the rounding error. We know that:
+	//
+	//   (target % types.SiafundCount) == (payout % types.SiafundCount)
+	//
+	// therefore, we can simply adjust the guess to have this remainder as
+	// well. The only wrinkle is that, since we know guess >= payout, if the
+	// guess remainder is smaller than the target remainder, we must subtract
+	// an extra types.SiafundCount.
+	//
+	// for example, if target = 87654321 and types.SiafundCount = 10000, then:
+	//
+	//   initial_guess  = 87654321 * (1 / (1 - tax))
+	//                  = 91211572
+	//   target % 10000 =     4321
+	//   adjusted_guess = 91204321
+
+	mod64 := func(c types.Currency, v uint64) types.Currency {
+		var r uint64
+		if c.Hi < v {
+			_, r = bits.Div64(c.Hi, c.Lo, v)
+		} else {
+			_, r = bits.Div64(0, c.Hi, v)
+			_, r = bits.Div64(r, c.Lo, v)
+		}
+		return types.NewCurrency64(r)
+	}
+	sfc := (consensus.State{}).SiafundCount()
+	tm := mod64(target, sfc)
+	gm := mod64(guess, sfc)
+	if gm.Cmp(tm) < 0 {
+		guess = guess.Sub(types.NewCurrency64(sfc))
+	}
+	return guess.Add(tm).Sub(gm)
+}
+
+func mineBlocks(log *zap.Logger, cm *chain.Manager, n int) {
+	for n > 0 {
+		b, ok := coreutils.MineBlock(cm, types.VoidAddress, 5*time.Second)
+		if !ok {
+			continue
+		} else if err := cm.AddBlocks([]types.Block{b}); err != nil {
+			log.Panic("failed to add funding block", zap.Error(err))
+		}
+		n--
+		log.Debug("mined block", zap.Stringer("index", cm.Tip()))
+	}
+}
+
+func setupV1Contracts(log *zap.Logger, cm *chain.Manager, pk types.PrivateKey, fundingBlock types.Block) FixtureV1ContractAddresses {
```

Comment:
Passing `fundingBlock` is a bit odd considering you have a wallet. I would just rely on the wallet to fund the contract rather than hacking together something using the miner payouts of a block directly.

It probably also makes sense to create the wallet outside of this function and then pass it to both setup functions unless you need different private keys to own these contracts.